text/3519-arbitrary-self-types-v2.md

```diff
@@ -0,0 +1,403 @@
+
+However, if it's OK to create a reference `&T`, you _probably don't need this feature_. You can probably simply use `&self` as your receiver type.
+
+This feature is fundamentally aimed at smart pointer types `P<T>` where it's not safe to create a reference `&T`. As noted above, that's most commonly because of semantic differences to pointers in other languages, but it might be because references have special meaning or behavior in some pure Rust domain. Either way, it's not OK to create a Rust reference `&T` or `&mut T`, yet we may want to allow methods to be called on some reference-like thing.
+
+For this reason, implementing `Deref::deref` is problematic for _nearly everyone who wants to use arbitrary self types_.
+
+If you're implementing a smart pointer `P<T>` yet you can't allow a reference `&T` to exist, any option for implementing `Deref::deref` has drawbacks:
+
+* Specify `Deref::Target=T` and panic in `Deref::deref`. Not good.
+* Specify `Deref::Target=*const T`. This works with the current arbitrary self types feature, but is only possible if your smart pointer type contains a `*const T` which you can reference - this isn't the case for (for instance) weak pointers or types containing `NonNull`.
+
+Therefore, the current Arbitrary Self Types v2 provides a separate `Receiver` trait.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When declaring a method, users can declare the type of the `self` receiver to be any type `T` where `T: Receiver<Target = Self>` or `Self`.
+
+The `Receiver` trait is simple and only requires to specify the `Target` type:
+
+```rust
+trait Receiver {
+    type Target: ?Sized;
+}
+```
+
+The `Receiver` trait is already implemented for many standard library types:
+- smart pointers in the standard library: `Rc<Self>`, `Arc<Self>`, `Box<Self>`, and `Pin<Ptr<Self>>` (and in fact, any type which implements `Deref`)
+- references: `&Self` and `&mut Self`
+- pointers: `*const Self` and `*mut Self`
+
+Shorthand exists for references, so that `self` with no ascription is of type `Self`, `&self` is of type `&Self` and `&mut self` is of type `&mut Self`.
+
+All of the following self types are valid:
+
+```rust
+impl Foo {
+    fn by_value(self /* self: Self */);
+    fn by_ref(&self /* self: &Self */);
+    fn by_ref_mut(&mut self /* self: &mut Self */);
+    fn by_ptr(self: *const Self);
+    fn by_mut_ptr(self: *mut Self);
+    fn by_box(self: Box<Self>);
+    fn by_rc(self: Rc<Self>);
+    fn by_custom_ptr(self: CustomPtr<Self>);
+}
+
+struct CustomPtr<T>(*const T);
+
+impl<T> Receiver for CustomPtr<T> {
+    type Target = T;
+}
+```
+
+## Recursive arbitrary receivers
+
+Receivers are recursive and therefore allowed to be nested. If type `T` implements `Receiver<Target=U>`, and type `U` implements `Receiver<Target=Self>`, `T` is a valid receiver (and so on outward).
```

Comment:
What is the motivation for this?

Also explain why is it done "outside" the trait -- otherwise we get overlapping instances.

Reply:
Hm. Good question. This recursiveness was a feature of the existing unstable `arbitrary_self_types` feature. The same overlappingness applies to `Deref` in just the same way as `Receiver` - we're not changing anything here.

For the inter-language interop use-cases, we don't have an immediate _need_ for the recursiveness, but my sense is that for other use cases it might be useful. The example below of `Rc<Box<Self>>` might be a good one, or `Pin<Box<Self>>`.

@clarfonthey I think you might be able to explain this better than me? :)

Reply:
This is already how the blessed subset works, so keeping this is required to get rid of the special-ness of the blessed std types 
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d12e5ac493a2b492da901c6b84c592c3
```rs
struct S;
impl S {
    fn f(self: std::pin::Pin<std::pin::Pin<&Self>>) {}
    fn f2(self: Box<Box<Self>>) {}
}
```

Reply:
Okay, fair. Would be good to explain that this is not new but just describing existing behavior.

Reply:
I put an example in another comment here describing how it works for `Deref`, using the example of a `Box<String>`: https://github.com/rust-lang/rfcs/pull/3519#discussion_r1380653042

Basically, because of how deref coercions work, the compiler can iteratively apply the deref (or in this case, mark a receiver), but a simple trait bound will only capture one iteration. This is definitely weird, but it's the status quo at least.

Reply:
Added a note that this is existing behavior.